ibtk/src/utilities/SecondaryHierarchy.cpp

```diff
@@ -40,19 +42,100 @@
+    {
+    }
+
+    virtual void resetHierarchyConfiguration(const tbox::Pointer<hier::BasePatchHierarchy<NDIM> > /*hierarchy*/,
+                                             const int /*coarsest_level*/,
+                                             const int /*finest_level*/) override
+    {
+    }
+
+    virtual void applyGradientDetector(const SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM> > hierarchy,
+                                       const int level_number,
+                                       const double /*error_data_time*/,
+                                       const int tag_index,
+                                       const bool /*initial_time*/,
+                                       const bool /*uses_richardson_extrapolation_too*/) override
+    {
+        if (level_number == hierarchy->getFinestLevelNumber()) return;
+        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy = hierarchy;
+        TBOX_ASSERT(patch_hierarchy);
+        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > level = hierarchy->getPatchLevel(level_number);
+
+        if (!level->checkAllocated(tag_index)) level->allocatePatchData(tag_index, 0.0);
+        HierarchyCellDataOpsInteger<NDIM> hier_cc_data_ops(hierarchy, level_number, level_number);
+        // this is the only time we will tag things so set the tag values to zero
+        hier_cc_data_ops.setToScalar(tag_index, 0);
+
+        // OK - now we need to figure out which values need to be refined. Since
+        // the hierarchy already satisfies a proper nesting condition we only
+        // need to check the next finer level to figure out which cells should
+        // be tagged on this level, since we don't care about tagging in the
+        // secondary hierarchy - just load balancing (but samrai requires we do
+        // both simultaneously)
+        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > finer_level = hierarchy->getPatchLevel(level_number + 1);
+        const auto& boxes = finer_level->getBoxes();
+        for (int i = 0; i < boxes.getNumberOfBoxes(); ++i)
+        {
+            const SAMRAI::hier::Box<NDIM> finer_box = boxes[i];
+            auto coarsened_box = finer_box;
+            coarsened_box.coarsen(finer_level->getRatioToCoarserLevel());
+            {
+                // Sanity check that we really do have proper nesting
+                auto check_box = coarsened_box;
+                check_box.refine(finer_level->getRatioToCoarserLevel());
+                TBOX_ASSERT(check_box == finer_box);
+            }
+
+            for (SAMRAI::hier::PatchLevel<NDIM>::Iterator p(level); p; p++)
+            {
+                const Pointer<Patch<NDIM> > patch = level->getPatch(p());
+                const Box<NDIM> box = patch->getBox();
+                Pointer<CellData<NDIM, int> > tag_data = patch->getPatchData(tag_index);
+
+                // SAMRAI's intersection code is wrong and returns nonsense
```

Comment:
Really??

Reply:
Unfortunately yes:
```cpp
template<int DIM> inline
Box<DIM> Box<DIM>::operator*(const Box<DIM>& box) const
{
   Box<DIM> both(*this);
   both.d_lo.max(box.d_lo);
   both.d_hi.min(box.d_hi);
   return(both);
}
```
this produces invalid boxes if the two boxes don't actually intersect. This causes some loops to behave very weirdly inside SAMRAI itself.

Reply:
I don't think SAMRAI has a notion of an empty box, but it should create one here.